### PR TITLE
Template Parts: Fix theme slug injection for patterns

### LINKED ISSUE #55202: Template Parts not displaying when inside patterns
## What problem does this address?
When a template part is inside a pattern block, it appears missing on the frontend. To reproduce:

1. Update to the latest version of WordPress trunk
2. Switch to the TT4 theme
3. Open the site in the frontend and you'll see this error:
```Template part has been deleted or is unavailable: header ```

## What is your proposed solution?
This seems to have been introduced in https://github.com/WordPress/gutenberg/pull/54651

The problem seems to be that although the template parts are given theme slugs when they are processed by `traverse_and_serialize_blocks`, when the block code itself comes to process the block the theme slug is missing. I can't work out why.

cc @gziolo as you worked on https://github.com/WordPress/gutenberg/pull/54651 :)

### PR DESCRIPTION
**Pull request by scruffian**
Fixes: https://github.com/WordPress/gutenberg/issues/55202
Trac ticket: https://core.trac.wordpress.org/ticket/59583

This ensures that the theme slug is correctly inserted into template parts that are inside patterns.

Furthermore, unit tests are added to prevent the same kind of regression in the future. The tests can be used to verify the fix: Cherry-pick 3cb17f8a1191924175ab8dc12c3d8e8394db91a4 and d077949adcb90db0d787e7136caa79765a6464f6 to `trunk`, run `npm run test:php -- --group=blocks`, and verify that it fails. Compare that to this branch where those tests pass (see CI, or test locally).

Props @gziolo for finding the problem

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by gziolo** [comment]
3 unit tests failed with these changes so that needs to be fixed before we proceed:

Screenshot 2023-10-10 at 14 13 10

**Comment by ockham** [comment]
Good catch, thank you both!

I'll try to write a quick unit test to cover the `theme` attribute insertion in patterns so we're guarded against this kind of regression in the future. I'll push a commit to this PR if that's okay.

**Comment by ockham** [comment]
For book-keeping, we should probably file a Trac ticket for this (since it needs to go into 6.4 Beta 3, and we're fairly late in the cycle). Would either of you mind doing so? Probably enough to basically copy-paste https://github.com/WordPress/gutenberg/issues/55202.

Otherwise I can take care of that once I have the unit test ready.

File: src/wp-includes/class-wp-block-patterns-registry.php [hunk 1]
```diff
@@ -164,12 +164,16 @@ public function unregister( $pattern_name ) {
...
+		$before_block_visitor = '_inject_theme_attribute_in_template_part_block';
+		$after_block_visitor  = null;
+		$hooked_blocks        = get_hooked_blocks();
 		if ( ! empty( $hooked_blocks ) || has_filter( 'hooked_block_types' ) ) {
-			$blocks               = parse_blocks( $content );
-			$before_block_visitor = make_before_block_visitor( $hooked_blocks, $pattern );
-			$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $pattern );
-			$content              = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+			$before_block_visitor = make_before_block_visitor( $hooked_blocks, $template );
+			$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $template );
```
Comment (ockham) [inline src/wp-includes/class-wp-block-patterns-registry.php]:
```suggestion
			$before_block_visitor = make_before_block_visitor( $hooked_blocks, $pattern );
			$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $pattern );
```

This should fix the failing unit tests.

Comment (ockham) [reply src/wp-includes/class-wp-block-patterns-registry.php]: Going to commit to get unit tests to pass.

**Comment by ockham** [comment]
> I'll try to write a quick unit test to cover the `theme` attribute insertion in patterns so we're guarded against this kind of regression in the future. I'll push a commit to this PR if that's okay.

https://github.com/WordPress/wordpress-develop/pull/5450/commits/3cb17f8a1191924175ab8dc12c3d8e8394db91a4

Verify that this covers the regression by cherry-picking the commit to `trunk`, and running `npm run test:php -- --group=blocks` there.

I'll add another test to cover `theme` injection in `get_all_registered`.

File: tests/phpunit/tests/blocks/wpBlockPatternsRegistry.php [hunk 2]
```diff
@@ -368,6 +368,29 @@ public function test_get_all_registered_includes_hooked_blocks() {
 		$this->assertSame( $expected, $registered );
 	}
 
+	/**
+	 * Should insert a theme attribute into Template Part blocks in registered patterns.
+	 *
+	 * @ticket XXXXX
```
Comment (ockham) [inline tests/phpunit/tests/blocks/wpBlockPatternsRegistry.php]: TBD

File: src/wp-includes/class-wp-block-patterns-registry.php [hunk 3]
```diff
@@ -164,12 +164,16 @@ public function unregister( $pattern_name ) {
 	 */
 	private function prepare_content( $pattern, $hooked_blocks ) {
 		$content = $pattern['content'];
+
+		$before_block_visitor = '_inject_theme_attribute_in_template_part_block';
+		$after_block_visitor  = null;
+		$hooked_blocks        = get_hooked_blocks();
```
Comment (ockham) [inline src/wp-includes/class-wp-block-patterns-registry.php]: No need for this line; it would negatively impact performance, so we should remove it.

Comment (ockham) [reply src/wp-includes/class-wp-block-patterns-registry.php]:
```suggestion
```

**Comment by ockham** [comment]
> I'll add another test to cover `theme` injection in `get_all_registered`.

https://github.com/WordPress/wordpress-develop/pull/5450/commits/d077949adcb90db0d787e7136caa79765a6464f6

Going to file the ticket so I can put the corred `@ticket` references in unit tests' PHPDoc.

**Review by gziolo — APPROVED** [review]
The code changes look good. New unit tests were added to cover the regression discovered. Let's ensure that all unit tests pass on CI.

**Comment by ockham** [comment]
Looking good. Thanks all, I'll land it 😊

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/56818.

**Pull request closed without merging**